### PR TITLE
Add parameter "source" to stack-analyses

### PR DESCRIPTION
**Pull request by samuzzal-choudhury**

### CODE REVIEW

File: bayesian/dependency_finder.py [hunk 1]
```diff
@@ -46,6 +48,14 @@ def execute(self, arguments, db, manifests):
         # If we receive a manifest file we need to save it first
         result = []
         for manifest in manifests:
+            resolve_poms = False
+            if source == 'osio':
+                content_hash = generate_content_hash(manifest['content'])
+                if pom_hash.get(content_hash) is not None:
+                    return {'result': pom_hash.get(content_hash)}
```
Comment (miteshvp) [inline bayesian/dependency_finder.py]: this early return will not work for multiple manifests

File: bayesian/dependency_finder.py [hunk 2]
```diff
@@ -104,4 +114,7 @@ def _flatten(deps, collect):
                 out["details"][0]['_resolved'] = resolved_deps
             result.append(out)
 
+        if source == 'osio':
```
Comment (miteshvp) [inline bayesian/dependency_finder.py]: storing of contents should be per manifest. This loop should be part of main for loop and not outside.

File: bayesian/dependency_finder.py [hunk 3]
```diff
@@ -104,4 +114,7 @@ def _flatten(deps, collect):
...
+            pom_hash[content_hash] = result
```
Comment (miteshvp) [inline bayesian/dependency_finder.py]: pom_hash[content_hash] = out

File: bayesian/dependency_finder.py (hunk 1 above)
Comment (samuzzal-choudhury) [reply bayesian/dependency_finder.py]: good catch

**Comment by samuzzal-choudhury** [comment]
[test]

File: bayesian/dependency_finder.py [hunk 4]
```diff
@@ -46,6 +49,15 @@ def execute(self, arguments, db, manifests):
...
         result = []
         for manifest in manifests:
+            content_hash = None
+            if source == 'osio':
+                content_hash = generate_content_hash(manifest['content'])
+                print("{} file digest is {}".format(manifest['filename'], content_hash))
+
+                s3 = AmazonS3(bucket_name='boosters-manifest')
+                s3.connect()
+                manifest['content'] = s3.retrieve_blob(content_hash).decode('utf-8')
```
Comment (miteshvp) [inline bayesian/dependency_finder.py]: we should handle error scenario as well

**Comment by tisnik** [comment]
@samuzzal-choudhury could you please provide unit tests for the added/changed code? TY in advance.

File: bayesian/dependency_finder.py [hunk 5]
```diff
@@ -46,6 +50,18 @@ def execute(self, arguments, db, manifests):
         # If we receive a manifest file we need to save it first
         result = []
         for manifest in manifests:
+            content_hash = None
+            if source == 'osio':
+                content_hash = generate_content_hash(manifest['content'])
+                print("{} file digest is {}".format(manifest['filename'], content_hash))
```
Comment (miteshvp) [inline bayesian/dependency_finder.py]: either we can use a logger error or we can `raise FatalTaskError`

File: bayesian/dependency_finder.py [hunk 6]
```diff
@@ -46,6 +50,18 @@ def execute(self, arguments, db, manifests):
...
+
+                s3 = AmazonS3(bucket_name='boosters-manifest')
+                try:
+                    s3.connect()
+                    manifest['content'] = s3.retrieve_blob(content_hash).decode('utf-8')
+                except ClientError as e:
+                    print("Unexpected error while retrieving S3 data: %s" % e)
```
Comment (miteshvp) [inline bayesian/dependency_finder.py]: either we can use a logger error or we can `raise FatalTaskError`

Comment (samuzzal-choudhury) [reply bayesian/dependency_finder.py]: @miteshvp this is within server. So, raising FatalTaskError does not make sense. Rather any exception here is going to send an HTTP 500 response.

**Comment by samuzzal-choudhury** [comment]
@tisnik please elaborate on the changes are you looking at?

**Comment by tisnik** [comment]
@samuzzal-choudhury there's new function in utils module and four new blocks of code that needed to be checked a bit.

**Comment by samuzzal-choudhury** [comment]
@tisnik sorry that the coverage results are going bad because of that. I can add the logic of the new function to the API code directly. But, the rest of the changes are necessary ones.

**Comment by tisnik** [comment]
@samuzzal-choudhury NP. actually IIRC all what's needed is one unit tests for the new function. Or -if you merge this PR, I can write the test later if you prefer.

**Comment by miteshvp** [comment]
@tisnik going to merge this to unblock others. Will be really  appreciated if you can add those tests as you suggested. TYVM in advance.
cc @samuzzal-choudhury

**Pull request merged**